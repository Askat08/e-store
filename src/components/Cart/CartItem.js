import React from "react";

export default function CartItem({ props, item }) {
  const { id, title } = item;
  const { path, extension } = item.thumbnail;
  const price = item.prices[0].price;
  const quantity = item.count;
  const https = Array.from(path);
  https.splice(4, 0, "s");
  return (
    <div className="row my-2 text-capitilize text-center">
      <div className="col-10 mx-auto col-lg-2">
        <img
          src={https.join("") + "." + extension}
          alt={title}
          className="img-fluid"
          style={{ width: "5rem", height: "5rem" }}
        />
      </div>
      <div className="col-10 mx-auto col-lg-2">
        <span className="d-lg-none">Comic : </span> {title}
      </div>
      <div className="col-10 mx-auto col-lg-2">
        <span className="d-lg-none">Price : </span>
        <strong>{`$${price.toFixed(2)}`}</strong>
      </div>
      <div className="col-10 mx-auto col-lg-2 my-2 my-lg-0">
        <div className="d-flex justify-content-center">
          <div className="">
            <span
              className="btn btn-cart mx-1"
              onClick={() => props.decrementCount(id)}
            >
              -
            </span>
            <span className="btn btn-cart mx-1">{quantity}</span>
            <span
              className="btn btn-cart mx-1"
              onClick={() => props.incrementCount(id)}
            >
              +
            </span>
          </div>
        </div>
      </div>

      <div className="col-10 mx-auto col-lg-2">
        <div
          style={{ cursor: "pointer" }}
          onClick={() => {
            props.removeItemFromCart(id, props.cart);
          }}
        >
          <i className="fas fa-trash text-danger" />
        </div>
      </div>

      <div className="col-10 mx-auto col-lg-2">
        <strong>Item total : $ {(price * quantity).toFixed(2)}</strong>
      </div>
    </div>
  );
}
