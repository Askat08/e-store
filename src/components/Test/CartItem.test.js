import React from "react";
import { shallow } from "enzyme";
import CartItem from "../Cart/CartItem";

const itemMock = {
  id: 23,
  title: "deadpool",
  thumbnail: { path: "http://file/path", extension: "jpg" },
  prices: [{ price: 2.99 }],
  count: 3,
};

const itemsArrayMock = [
  { title: "iron man", id: 5, count: 1, prices: [{ price: 2.99 }] },
  { title: "deadpool", id: 6, count: 1, prices: [{ price: 2.25 }] },
  { title: "daredevil", id: 7, count: 1, prices: [{ price: 23.99 }] },
  { title: "captain America", id: 8, count: 1, prices: [{ price: 15.19 }] },
];

const propsMock = {
  decrementCount: () => (itemMock.count = itemMock.count - 1),
  incrementCount: () => (itemMock.count = itemMock.count + 1),
  removeItemFromCart: () => itemsArrayMock.pop(),
};

describe("CartItem component test", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<CartItem item={itemMock} props={propsMock} />);
  });

  it("should render CartItem component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("all div's has proper props", () => {
    expect(
      wrapper.find("div").at(0).hasClass("row my-2 text-capitilize text-center")
    ).toBeTruthy();
    expect(wrapper.find("div").at(1).props().className).toEqual(
      "col-10 mx-auto col-lg-2"
    );
    expect(
      wrapper.find("div").at(2).hasClass("col-10 mx-auto col-lg-2")
    ).toBeTruthy();
    expect(
      wrapper.find("div").at(3).hasClass("col-10 mx-auto col-lg-2")
    ).toBeTruthy();
    expect(
      wrapper.find("div").at(4).hasClass("col-10 mx-auto col-lg-2 my-2 my-lg-0")
    ).toBeTruthy();
    expect(
      wrapper.find("div").at(5).hasClass("d-flex justify-content-center")
    ).toBeTruthy();
    expect(
      wrapper.find("div").at(7).hasClass("col-10 mx-auto col-lg-2")
    ).toBeTruthy();
    expect(wrapper.find("div").at(8).props().style).toEqual(expect.any(Object));
    expect(wrapper.find("div").at(8).props().onClick).toEqual(
      expect.any(Function)
    );
    expect(
      wrapper.find("div").at(9).hasClass("col-10 mx-auto col-lg-2")
    ).toBeTruthy();
  });

  it("image should have all props", () => {
    expect(wrapper.find("img").props()).toEqual({
      className: "img-fluid",
      src: expect.any(String),
      alt: itemMock.title,
      style: {
        height: "5rem",
        width: "5rem",
      },
    });
  });

  it("span should have all props", () => {
    expect(wrapper.find("span").at(0).hasClass("d-lg-none")).toBeTruthy();
    expect(wrapper.find("span").at(0).text()).toEqual("Comic : ");
    expect(wrapper.find("span").at(1).hasClass("d-lg-none")).toBeTruthy();
    expect(wrapper.find("span").at(1).text()).toEqual("Price : ");
    expect(
      wrapper.find("span").at(2).hasClass("btn btn-cart mx-1")
    ).toBeTruthy();
    expect(wrapper.find("span").at(2).props().onClick).toEqual(
      expect.any(Function)
    );
    expect(wrapper.find("span").at(2).text()).toEqual("-");
    expect(
      wrapper.find("span").at(3).hasClass("btn btn-cart mx-1")
    ).toBeTruthy();
    expect(
      wrapper.find("span").at(4).hasClass("btn btn-cart mx-1")
    ).toBeTruthy();
    expect(wrapper.find("span").at(4).props().onClick).toEqual(
      expect.any(Function)
    );
    expect(wrapper.find("span").at(4).text()).toEqual("+");
  });

  it("strong should have contain proper text", () => {
    expect(wrapper.find("strong").at(0).text()).toEqual(
      "$" + itemMock.prices[0].price
    );
    expect(wrapper.find("strong").at(1).text()).toEqual(
      "Item total : $ " + (itemMock.prices[0].price * itemMock.count).toFixed(2)
    );
  });

  it("onClick method should call callback", () => {
    const item = {
      title: "iron man",
      id: 5,
      count: 1,
      prices: [{ price: 2.99 }],
    };
    wrapper.find("span").at(2).props().onClick(); // decrement count
    expect(itemMock.count).toEqual(2);
    wrapper.find("span").at(4).props().onClick(); // increment count
    expect(itemMock.count).toEqual(3);
    wrapper.find("div").at(8).props().onClick(); // remove item
    expect(itemsArrayMock.length).toEqual(3);
    wrapper.find("div").at(8).props().onClick(); // remove item
    expect(itemsArrayMock.length).toEqual(2);
  });
});
