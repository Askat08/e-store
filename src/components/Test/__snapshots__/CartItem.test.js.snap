// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem component test should render CartItem component 1`] = `
<div
  className="row my-2 text-capitilize text-center"
>
  <div
    className="col-10 mx-auto col-lg-2"
  >
    <img
      alt="deadpool"
      className="img-fluid"
      src="https://file/path.jpg"
      style={
        Object {
          "height": "5rem",
          "width": "5rem",
        }
      }
    />
  </div>
  <div
    className="col-10 mx-auto col-lg-2"
  >
    <span
      className="d-lg-none"
    >
      Comic : 
    </span>
     
    deadpool
  </div>
  <div
    className="col-10 mx-auto col-lg-2"
  >
    <span
      className="d-lg-none"
    >
      Price : 
    </span>
    <strong>
      $2.99
    </strong>
  </div>
  <div
    className="col-10 mx-auto col-lg-2 my-2 my-lg-0"
  >
    <div
      className="d-flex justify-content-center"
    >
      <div
        className=""
      >
        <span
          className="btn btn-cart mx-1"
          onClick={[Function]}
        >
          -
        </span>
        <span
          className="btn btn-cart mx-1"
        >
          3
        </span>
        <span
          className="btn btn-cart mx-1"
          onClick={[Function]}
        >
          +
        </span>
      </div>
    </div>
  </div>
  <div
    className="col-10 mx-auto col-lg-2"
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <i
        className="fas fa-trash text-danger"
      />
    </div>
  </div>
  <div
    className="col-10 mx-auto col-lg-2"
  >
    <strong>
      Item total : $ 
      8.97
    </strong>
  </div>
</div>
`;
